Use bound theme properties for documentation

YuriSizov opened the pull request:
Based on https://github.com/godotengine/godot/pull/81551 ~(so drafting for now)~.

This PR makes use of the new theme item binding information for documentation generation. This means that the decision which properties to document no longer comes from what is defined with the default engine theme. Instead, just like with properties and methods, only what's registered by the class itself (and it's inherited classes) is added.

As a result, a couple of theme properties were removed since they aren't actually used by their respective classes (maybe they should, but for now they aren't, so removing them is a better option).

One open question is what to do with inherited theme properties. Historically, due to our reliance on the default theme, all inherited properties would be displayed in extending classes as if their own. This is handy, since you can always see every property that can affect the GUI node, but that's not how we do the docs for class members, signals, and methods. We can, now, filter these out and remove them from extending classes or add them as stubs without descriptions like we do for "overridden" class members.

For now I'm going with adding them everywhere. This is most consistent with our current documentation, for better or worse. We can reconsider this in a future PR, I think.

PS. Similar PRs will be made for theme overrides, code completion, and the theme editor. Going one by one to slowly figure out if anything is missing from the API (like this PR added data types to the bind information).

KoBeWi commented on doc/classes/CodeEdit.xml:
```diff
@@ -654,9 +654,6 @@
 		</theme_item>
-		<theme_item name="completion_font_color" data_type="color" type="Color" default="Color(0.67, 0.67, 0.67, 1)">
-			Font [Color] for the code completion popup.
-		</theme_item>
```
What happened to this color?

KoBeWi commented:
I think displaying every inherited theme property is counter-productive. The documentation ends up with tens of new useless (and empty!) items. I'm not even sure if all of them are functional (some are though, I checked). I think it's better if they did not appear at all, like regular member, and in rare cases when the item's usage differ, it can be noted in the class description (I can only think of ColorPickerButton, which hides most of stuff).

YuriSizov commented:
> Also entries that got removed should be removed from default theme too I think.

Did I miss any? I removed them from both the default theme and the editor theme.

KoBeWi commented:
No, I missed your theme changes 🤦‍♂️

akien-mga commented:
> One open question is what to do with inherited theme properties. Historically, due to our reliance on the default theme, all inherited properties would be displayed in extending classes as if their own. This is handy, since you can always see every property that can affect the GUI node, but that's not how we do the docs for class members, signals, and methods. We can, now, filter these out and remove them from extending classes or add them as stubs without descriptions like we do for "overridden" class members.

I think being consistent with regular properties, etc. would be good. The big problem with duplicating them everywhere is that we need to duplicate their documentation everywhere too. So I'd prefer they're documented only in the base class (and overrides are shown like we do for properties).

If we still want a way to easily give an overview over all the theme properties a given class has access to, that's an orthogonal UX problem IMO, we can figure out solutions that don't require duplicating the XML and documentation data. (E.g. a folded-by-default list of all theme properties, below the list of theme properties added in this class and their documentation.)

YuriSizov commented:
Alright, I removed inherited properties from documentation and unified some descriptions for H/V nodes. Inspector still works correctly and shows descriptions and redirects to correct pages. So I think we're good to go.

YuriSizov committed:
Use bound theme properties for documentation

akien-mga commented:
You need to copy back the descriptions to the base classes where the theme properties are now being exposed. We seem to be losing quite a bit of documentation otherwise.

KoBeWi commented:
The "new" empty items are properties that weren't documented before and they never had description. The descriptions in inherited classes were copy-paste from the base class (with some changed references).

YuriSizov commented:
Yes, I copied stuff that was only defined in extending classes or had differences not reflected in the parent class (like H/V variants). Maybe I missed some? But most of the descriptions that are now gone are duplicates of the base class documentation, as @KoBeWi says. Like in the case of a dozen Button types.

KoBeWi commented on scene/theme/theme_db.cpp:
```diff
@@ -370,6 +371,19 @@ void ThemeDB::update_class_instance_items(Node *p_instance) {
+	HashMap<StringName, HashMap<StringName, ThemeItemBind>>::Iterator E = theme_item_binds.find(p_class_name);
+	if (E) {
+		for (const KeyValue<StringName, ThemeItemBind> &F : E->value) {
+			if (F.value.external || F.value.class_name != p_class_name) {
```
When the class name can be different if the item is not external?

YuriSizov replied:
For the moment it can't, but this permutation is possible so I thought I'd add a check.

YuriSizov replied:
I actually need to rework this method, and I will need to use this check, because I'm going to sort through inherited classes.

KoBeWi commented on doc/classes/Button.xml:
```diff
@@ -131,20 +131,34 @@
 		<theme_item name="disabled" data_type="style" type="StyleBox">
 			[StyleBox] used when the [Button] is disabled.
 		</theme_item>
+		<theme_item name="disabled_mirrored" data_type="style" type="StyleBox">
```
This one has description in OptionButton.

KoBeWi commented on doc/classes/Button.xml:
```diff
@@ -131,20 +131,34 @@
 		</theme_item>
+		<theme_item name="hover_mirrored" data_type="style" type="StyleBox">
+		</theme_item>
+		<theme_item name="hover_pressed" data_type="style" type="StyleBox">
```
This one has description in CheckBox and CheckButton.

KoBeWi commented on doc/classes/GraphElement.xml:
```diff
@@ -66,4 +66,8 @@
 		</signal>
 	</signals>
+	<theme_items>
+		<theme_item name="resizer" data_type="icon" type="Texture2D">
```
This one has description in GraphNode.

YuriSizov replied:
Wait, that's weird. It should've been removed from `GraphNode`.

YuriSizov replied:
Ah! I think there are going to be a few more duplicated items to remove.

KoBeWi commented on doc/classes/LinkButton.xml:
```diff
@@ -62,12 +62,16 @@
 		<theme_item name="font_color" data_type="color" type="Color" default="Color(0.875, 0.875, 0.875, 1)">
 			Default text [Color] of the [LinkButton].
 		</theme_item>
+		<theme_item name="font_disabled_color" data_type="color" type="Color" default="Color(0, 0, 0, 1)">
```
This one has description in all Button classes (LinkButton is BaseButton).

KoBeWi reviewed (approved):
Looks fine, but I left some comments for theme item descriptions that you could copy from the deleted ones.

YuriSizov commented:
Added missed descriptions, and modified `ThemeDB::get_class_own_items` so inherited properties which are cached by extending classes are also properly ignored. This removed a bunch more theme items. I added missing descriptions resulting from that too.

akien-mga commented:
Thanks!
